dm/syncer/data_validator.go

```diff
@@ -698,69 +715,58 @@ func (v *DataValidator) processRowsEvent(header *replication.EventHeader, ev *re
+		switch changeType {
+		case rowInsert:
+			afterImage = ev.Rows[i]
+		case rowUpdated:
+			beforeImage, afterImage = ev.Rows[i], ev.Rows[i+1]
+		default: // rowDeleted
+			beforeImage = ev.Rows[i]
 		}
-		key := genRowKey(pkValue)
-
-		if changeType == rowUpdated {
-			afterRowChangeType := changeType
-			afterRow := ev.Rows[i+1]
-			afterPkValue := make([]string, len(pkCols))
-			for idx, col := range pkCols {
-				afterPkValue[idx] = genColData(afterRow[col.Offset])
-			}
-			afterKey := genRowKey(afterPkValue)
-			if afterKey != key {
-				// TODO: may reuse IsIdentityUpdated/SplitUpdate of RowChange in pkg/sqlmodel
-				v.dispatchRowChange(key, &rowChange{
-					table: table,
-					Key:   key,
-					Data:  row,
-					Tp:    rowDeleted,
-				})
-				v.processedRowCounts[rowDeleted].Inc()
-				afterRowChangeType = rowInsert
-			}
-			v.dispatchRowChange(afterKey, &rowChange{
-				table: table,
-				Key:   afterKey,
-				Data:  afterRow,
-				Tp:    afterRowChangeType,
-			})
-			v.processedRowCounts[afterRowChangeType].Inc()
+
+		rowChange := sqlmodel.NewRowChange(
+			&cdcmodel.TableName{Schema: sourceTable.Schema, Table: sourceTable.Name},
+			&cdcmodel.TableName{Schema: targetTable.Schema, Table: targetTable.Name},
+			beforeImage, afterImage,
+			tableInfo, downstreamTableInfo.TableInfo,
+			nil,
+		)
+		rowChange.SetWhereHandle(downstreamTableInfo.WhereHandle)
+		if changeType == rowUpdated && rowChange.IsIdentityUpdated() {
+			delRow, insRow := rowChange.SplitUpdate()
+			delRowKey := genRowKey(delRow)
+			v.dispatchRowChange(delRowKey, &rowChangeJob{Key: delRowKey, Tp: rowDeleted, row: delRow})
+			v.processedRowCounts[rowDeleted].Inc()
+
+			insRowKey := genRowKey(insRow)
+			v.dispatchRowChange(insRowKey, &rowChangeJob{Key: insRowKey, Tp: rowInsert, row: insRow})
+			v.processedRowCounts[rowInsert].Inc()
 		} else {
-			v.dispatchRowChange(key, &rowChange{
-				table: table,
-				Key:   key,
-				Data:  row,
-				Tp:    changeType,
-			})
+			rowKey := genRowKey(rowChange)
+			v.dispatchRowChange(rowKey, &rowChangeJob{Key: rowKey, Tp: changeType, row: rowChange})
 			v.processedRowCounts[changeType].Inc()
 		}
 	}
 	return nil
 }
 
-func (v *DataValidator) checkAndFlushCheckpoint(loc binlog.Location) error {
+func (v *DataValidator) checkAndPersistCheckpointAndData(loc binlog.Location) error {
 	metaFlushInterval := v.cfg.ValidatorCfg.MetaFlushInterval.Duration
 	if time.Since(v.lastFlushTime) > metaFlushInterval {
 		v.lastFlushTime = time.Now()
-		if err := v.flushCheckpointAndData(loc); err != nil {
+		if err := v.persistCheckpointAndData(loc); err != nil {
 			v.L.Warn("failed to flush checkpoint: ", zap.Error(err))
 			return err
 		}
 	}
 	return nil
 }
 
-func (v *DataValidator) flushCheckpointAndData(loc binlog.Location) error {
+func (v *DataValidator) persistCheckpointAndData(loc binlog.Location) error {
 	var wg sync.WaitGroup
 	wg.Add(v.workerCnt)
-	flushJob := &rowChange{
+	flushJob := &rowChangeJob{
```

Comment:
Little complaint: a bit weird to name a flushJob `rowChangeJob`. It seems these jobs are all assigned to workers so it'd be better called `workerJob` (or something like that). No need to change, I think, because it would be a big work to do. I'm happy to leave it here😂

Reply:
renaming is easy using goland.